### DISCUSSION TITLE
Ability to disable or overwrite behavior for control tooltips

### DISCUSSION THREAD

**Discussion by LunaCapra** [godotengine/godot-proposals#8858] [GUI]
I'm making a game where I would like for certain control nodes to have tooltips, however, I want to display these tooltips in a completely custom way, rather than the preset `PopupPanel` and I think it would be nice to be able to use the pre-existing variable for a tooltip text rather than creating a redundant second tooltip text variable in a script.

**Comment by Calinou** [comment]
> and I think it would be nice to be able to use the pre-existing variable for a tooltip text rather than creating a redundant second tooltip text variable in a script.

You should already be able to use the existing `tooltip_text` property in your custom tooltip.

**Reply by LunaCapra** [reply]
Correct, but the `PopupPanel` shows up when hovering over the control with a mouse.
I can use `_make_custom_tooltip` to show an empty tooltip, but the panel behind it will still remain and I do want the mouse to still cause a tooltip to appear, so setting the mouse focus to ignore won't work.

**Comment by LunaCapra** [comment]
To expand a bit on the idea:

The problem for me is that the tooltip showing up being dependent on the text being empty or not is not very flexible.
When overwriting the existing `_make_custom_tooltip`, it only replaces the content *inside* the `PopupPanel`, without being able to modify the panel itself.

I also want to note that returning `null` from the `_make_custom_tooltip` function just results in the default behavior, rather than disabling the tooltip (Which I would have found to be much more intuitive and useful)

I have a few suggestions for this:

1. Keep the default behavior unless `_make_custom_tooltip` is overridden via script. If it is overridden, make returning `null` disable the tooltip completely, otherwise always show it (This would likely also address the confusion for https://github.com/godotengine/godot-proposals/issues/2169)
2. Separate the "showing the tooltip" functionality into a boolean setting that is on by default, rather than tying it to the text being empty or not. This way the user can additionally control whether the tooltip shows up on hover automatically or not without having to attach a custom script to every control node that needs it.
This could potentially optionally also propagate to all children of the control that has it disabled but this isn't entirely necessary.

Alternatively, the `_make_custom_tooltip` function could be changed or another function added so that you can show (or not show) something where the engine determines the tooltip should show up.
E.g. `_show_tooltip(at_position : Vector2)`
In turn letting the user decide entirely how to show the tooltip if needed.

**Reply by GopherDevGames** [reply]
I think the tooltip logic also takes into account the elements size, so a `_show_tooltip(at_position : Vector2)`, wouldn't have information about that, meaning that your tooltip could end up off-screen. Wouldn't a general `get_tooltip_position(size :Vector2)` be better? Being able to get called whenever and would allow the user to specify the size so the returned position could make sure it ends up fully on screen. Could also be paired  with `_make_custom_tooltip()` to spawn it at an offset etc.

**Reply by LunaCapra** [reply]
That would probably be better, yes.
This could also potentially be expanded to something more general rather than tooltips
e.g. `get_on_screen_position(size : Vector2)` or `get_position_in_bounds(size : Vector2, bounds : Rect2)` etc.

Maybe even as a function on `Rect2`

**Comment by voidexp** [comment]
It would be cool to support this on Viewport level, as a checkbox. Especially useful for making highly stylized 2D interfaces in 3D environments (like futuristic touch-interface computers), where the popups are out of place and annoying.

**Comment by Calinou** [comment]
Thanks for the proposal! Consolidating in https://github.com/godotengine/godot-proposals/issues/13325.
